kratos/utilities/auxiliary_global_master_slave_relation.h

```diff
@@ -0,0 +1,273 @@
+    KRATOS_CLASS_POINTER_DEFINITION(AuxiliaryGlobalMasterSlaveRelation);
+
+    /**
+     * @brief Constructor of the class
+     * @param SlaveEquationId the slave equation id for which this class is being constructed.
+     */
+    explicit AuxiliaryGlobalMasterSlaveRelation(IndexType SlaveEquationId = 0) : IndexedObject(SlaveEquationId), mConstant(0.0)
+    {
+        mLhsValue = 0.;
+        mRhsValue = 0.;
+        mConstant = 0.;
+    }
+
+    /**
+     * @brief Function to get the slave equation Id corresponding to this constraint.
+     * @param Constant the value of the constant to be assigned.
+     */
+    IndexType SlaveEquationId() const { return this->Id(); }
+
+    /**
+     * @brief Function to set the lefthand side of the constraint (the slave dof value)
+     * @param LhsValue the value of the lhs (the slave dof value)
+     */
+    void SetLeftHandSide(const double LhsValue)
+    {
+        mLockObject.SetLock();
+            mLhsValue = LhsValue;
+        mLockObject.UnSetLock();
+    }
+
+    /**
+     * @brief Function to update the righthand side of the constraint (the combination of all the master dof values and constants)
+     * @param RHSValue the value of the lhs (the slave dof value)
+     */
+    void SetRightHandSide(const double RhsValue)
+    {
+        mRhsValue = RhsValue;
+    }
+    void UpdateRightHandSide(const double RhsValueUpdate)
+    {
+        mLockObject.SetLock();
+            mRhsValue = mRhsValue + RhsValueUpdate;
+        mLockObject.UnSetLock();
+    }
+
+    // Get number of masters for this slave
+    IndexType NumberOfMasters() const
+    {
+        return mMasterEquationIdVector.size();
+    }
+
+    /**
+     * @brief this determines the master equation IDs connected to this constraint
+     * @param rResult the elemental equation ID vector
+     */
+    virtual void EquationIdsVector(IndexType& rSlaveEquationId,
+                                  EquationIdVectorType& rMasterEquationIds)
+    {
+        if (rMasterEquationIds.size() == 0)
+            rMasterEquationIds.resize(this->NumberOfMasters(), false);
+
+        rSlaveEquationId = this->SlaveEquationId();
+        rMasterEquationIds = mMasterEquationIdVector;
+    }
+
+    /**
+     * @brief   this is called during the assembling process in order
+     *          to calculate all elemental contributions to the global system
+     * matrix and the right hand side
+     * @param rMasterWeightsVector the elemental left hand side matrix
+     * @param rConstant the elemental right hand side
+     */
+    virtual void CalculateLocalSystem(VectorType &rMasterWeightsVector,
+                                      double &rConstant)
+    {
+        if (rMasterWeightsVector.size() == 0)
+            rMasterWeightsVector.resize(this->NumberOfMasters(), false);
+
+        for (IndexType i = 0; i < this->NumberOfMasters(); ++i)
+            rMasterWeightsVector(i) = mMasterWeightsVector[i];
+
+
+        mLockObject.SetLock();
```

Comment:
can't we do this with an openmp atomic?

      double tmp = mRhsValue - mLhsValue
      #pragma omp atomic
      rConstant = tmp;

that would also allow to avoid storing mConstant...

Reply:
mmm, i correct myself. guess this would imply OpenMP 3+ and hence not be available in windows.

nevertheless this part of the code still gives me a bad feeling. can u expand a bit on what you are trying to do here? ... i mean, why this is strictly needed...

Reply:
Here this is required because, when in the builder and solver , we are actually imposing the constraint on the update of the DOF value (residual formulation), this does not necessarily guarantee the DOFs themselves follow the constraint equation. So, we calculate the LHS value and RHS value of the constraint equation (with DOF values) and if they are not satisfying the constraint, we use the residual (of the constraint equation) as the constant.